src/Fixer/Casing/MagicMethodCasingFixer.php

```diff
@@ -0,0 +1,226 @@
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author SpacePossum
+ */
+final class MagicMethodCasingFixer extends AbstractFixer
+{
+    private static $magicNames = [
+        '__call' => '__call',
+        '__callstatic' => '__callStatic',
+        '__clone' => '__clone',
+        '__construct' => '__construct',
+        '__debuginfo' => '__debugInfo',
+        '__destruct' => '__destruct',
+        '__get' => '__get',
+        '__invoke' => '__invoke',
+        '__isset' => '__isset',
+        '__set' => '__set',
+        '__set_state' => '__set_state',
+        '__sleep' => '__sleep',
+        '__tostring' => '__toString',
+        '__unset' => '__unset',
+        '__wakeup' => '__wakeup',
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Magic method definitions and calls must be using the correct casing.',
+            [
+                new CodeSample(
+'<?php
+class Foo
+{
+    public function __Sleep()
+    {
+    }
+}
+'
+                ),
+                new CodeSample(
+'<?php
+$foo->__INVOKE(1);
+'
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isTokenKindFound(T_STRING);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $inClass = 0;
+        $tokenCount = count($tokens);
+
+        for ($index = 1; $index < $tokenCount - 2; ++$index) {
+            if (0 === $inClass && $tokens[$index]->isClassy()) {
+                $inClass = 1;
+                $index = $tokens->getNextTokenOfKind($index, ['{']);
+
+                continue;
+            }
+
+            if (0 !== $inClass) {
+                if ($tokens[$index]->equals('{')) {
+                    ++$inClass;
+
+                    continue;
+                }
+
+                if ($tokens[$index]->equals('}')) {
+                    --$inClass;
+
+                    continue;
+                }
+            }
+
+            if (!$tokens[$index]->isGivenKind(T_STRING)) {
+                continue; // wrong type
+            }
+
+            $content = $tokens[$index]->getContent();
+            if ('__' !== substr($content, 0, 2)) {
+                continue; // cheap look ahead
+            }
+
+            $name = strtolower($content);
+
+            if (!$this->isMagicMethodName($name)) {
+                continue; // method name is not one of the magic ones we can fix
+            }
+
+            $nameInCorrectCasing = $this->getMagicMethodNameInCorrectCasing($name);
+            if ($nameInCorrectCasing === $content) {
+                continue; // method name is already in the correct casing, no fix needed
+            }
+
+            if ($this->isMethodSignature($tokens, $index)) {
+                if (0 !== $inClass) {
```

Comment:
This is misleading: the current index is a method signature, as just checked before, so the current index *is* in a class. The method name is actually wrong and should be `isFunctionSignature` IMO.

Reply:
Yes, but this can be misleading when you don't know what the code does yet. The more the code is clear about that, the easier to understand it and maintain it. The first thing I asked myself was why is the second "are we in a class" check required? Then I remembered the previous condition actually doesn't check that.

Reply:
>The more the code is clear about that, the easier to understand it and maintain it.

I understand that concept

>why is the second "are we in a class" check required?

It isn't, however if we don't do that global functions named the same a magic methods get fixed as well ( even though it would change code behavior) and you asked me to update the fixer not to do that.

Reply:
I've renamed the method and added a comment in the code, lemme know if this is sufficient.

Reply:
It is, thanks :) Just a small typo: _**an** method_ => _**a** method_.

Reply:
fixed, thanks :)